numpy/lib/function_base.py

```diff
@@ -2267,18 +2269,36 @@ class vectorize:
            [0., 0., 1., 2., 1., 0.],
            [0., 0., 0., 1., 2., 1.]])
 
+    Decorator syntax is supported.  The decorator can be called as
+    a function to provide keyword arguments.
+    >>>@np.vectorize
+    ...def identity(x):
+    ...    return x
+    ...
+    >>>identity([0, 1, 2])
+    array([0, 1, 2])
+    >>>@np.vectorize(otypes=[float])
+    ...def as_float(x):
+    ...    return x
+    ...
+    >>>as_float([0, 1, 2])
+    array([0., 1., 2.])
     """
-    def __init__(self, pyfunc, otypes=None, doc=None, excluded=None,
-                 cache=False, signature=None):
+    def __init__(self, pyfunc=np._NoValue, otypes=None, doc=None,
+                 excluded=None, cache=False, signature=None):
```

Comment:
What happens if the decorator is used with positional args
```
@vectorize(['float64'], "return float absolute value")
def myabs(a):
    return abs(a)
```

vs.
```
myabs = np.vectorize(abs, ['float64'], "return float absolute value")
```

I don't think we can change the signature to require kwargs since that would break existing code
```
def __init__(self, *, pyfunc=np._NoValue, ...
```

Reply:
>What happens if the decorator is used with positional args

Doesn't work, it assigns the list to pyfunc. Not sure how this can be fixed. Type annotations might work, but I feel like there's a better solution.

Reply:
One could check whether the first argument is `callable`? But it would then mean shifting the rest. Perhaps it is OK to insist on keyword arguments if used as a decorator?

Reply:
> But it would then mean shifting the rest

This is what I went for. It solves the issue, but it requires a hideous sequence of nested if-else statements, although it's oddly pep8-compliant.

Reply:
My gut feeling would be to just bail and raise an informative error when `pfunc` is not callable.  Allowing `None` in that sense might even be useful, but I don't care, the error could say: if you are using a decorator, you must use kwargs.

That said, if you all like it, its fine.  Just such logic tends to be hairy...

There may be a point on introducing e.g. `vectorized` as a new name or so and trying to move users over, but while I would like it, I feel it should probably come with deeper improvements of `vectorize`.

Reply:
>My gut feeling would be to just bail and raise an informative error when pfunc is not callable. Allowing None in that sense might even be useful, but I don't care, the error could say: if you are using a decorator, you must use kwargs.

I'd do that if there was a consensus about it, but I'd prefer not to. Generally, I don't like to shoehorn the user into doing things like having to specify the keywords in arguments. It'd be inconvenient for the user and likely cause bugs because of users misspelling keywords.

Reply:
I'm with @seberg here -- and keyword arguments would definitely seem better when used inside code, as a decorator, where really one has to worry about readability first.

Reply:
Ok, I changed it.